optaweb-vehicle-routing-backend/src/test/java/org/optaweb/vehiclerouting/plugin/planner/SolverManagerTest.java

```diff
@@ -0,0 +1,218 @@
+    void startSolver_should_start_solver() {
+        returnSolverFutureWhenSolverIsStarted();
+        solverManager.startSolver(mock(VehicleRoutingSolution.class));
+        verify(solver).solve(any(VehicleRoutingSolution.class));
+
+        // cannot start solver that is already solving
+        assertThatIllegalStateException()
+                .isThrownBy(() -> solverManager.startSolver(mock(VehicleRoutingSolution.class)));
+    }
+
+    @Test
+    void stopSolver_should_terminate_solver() {
+        returnSolverFutureWhenSolverIsStarted();
+        solverManager.startSolver(mock(VehicleRoutingSolution.class));
+        solverManager.stopSolver();
+        verify(solver).terminateEarly();
+
+        // another stopSolver() does nothing
+        solverManager.stopSolver();
+        // This verifies there were no more invocations of terminateEarly() without clearing all invocations.
+        // Not using Mockito.clearInvocations() only because it doesn't like generic arguments.
+        verify(solver).terminateEarly();
+    }
+
+    @Test
+    void reset_interrupted_flag() throws ExecutionException, InterruptedException {
+        returnSolverFutureWhenSolverIsStarted();
+        // start solver
+        solverManager.startSolver(mock(VehicleRoutingSolution.class));
+        when(solverFuture.isDone()).thenReturn(true);
+        when(solverFuture.get()).thenThrow(InterruptedException.class);
+
+        assertThatExceptionOfType(RuntimeException.class)
+                .isThrownBy(() -> solverManager.removeLocation(new RoadLocation()));
+        assertThat(Thread.interrupted()).isTrue();
+
+        assertThatExceptionOfType(RuntimeException.class)
+                .isThrownBy(() -> solverManager.stopSolver());
+        assertThat(Thread.interrupted()).isTrue();
+    }
+
+    @Test
+    void change_operations_should_fail_if_solver_has_not_started_yet() {
+        assertThatIllegalStateException()
+                .isThrownBy(() -> solverManager.addVehicle(mock(Vehicle.class)))
+                .withMessageContaining("started");
+        assertThatIllegalStateException()
+                .isThrownBy(() -> solverManager.removeVehicle(mock(Vehicle.class)))
+                .withMessageContaining("started");
+        assertThatIllegalStateException()
+                .isThrownBy(() -> solverManager.addLocation(mock(RoadLocation.class)))
+                .withMessageContaining("started");
+        assertThatIllegalStateException()
+                .isThrownBy(() -> solverManager.removeLocation(mock(RoadLocation.class)))
+                .withMessageContaining("started");
+    }
+
+    @Test
+    void change_operations_should_fail_is_solver_has_died() throws ExecutionException, InterruptedException {
+        returnSolverFutureWhenSolverIsStarted();
+        solverManager.startSolver(mock(VehicleRoutingSolution.class));
+        when(solverFuture.isDone()).thenReturn(true);
+        when(solverFuture.get()).thenThrow(ExecutionException.class);
```

Comment:
This test is great, still I have an idea how to improve it a bit:
- let's throw an instance of `ExecutionException` (see the overloaded `thenThrow()`) containing the cause
- assert also the cause

Reply:
This is already covered by `SolverExceptionTest`.